pkg/virt-launcher/virtwrap/device/sriov/hostdev.go

```diff
@@ -82,3 +90,88 @@ func createHostDevice(iface v1.Interface, hostPCIAddress string) (*api.HostDevic
 
 	return hostDev, nil
 }
+
+type deviceDetacher interface {
+	DetachDevice(xmlData string) error
+}
+
+func DetachHostDevices(dom deviceDetacher, hostDevices []api.HostDevice) error {
+	for _, device := range hostDevices {
+		devXML, err := xml.Marshal(device)
+		if err != nil {
+			return fmt.Errorf("failed to encode (xml) device %v, err: %v", device, err)
+		}
+		err = dom.DetachDevice(string(devXML))
+		if err != nil {
+			return fmt.Errorf("failed to detach device %s, err: %v", devXML, err)
+		}
+		log.Log.Infof("Successfully hot-unplug device: %s (%v)", device.Alias.Name, device.Source.Address)
+	}
+	return nil
+}
+
+func FilterHostDevices(domainSpec *api.DomainSpec) []api.HostDevice {
+	var hostDevices []api.HostDevice
+
+	for _, device := range domainSpec.Devices.HostDevices {
+		if device.Alias != nil && strings.HasPrefix(device.Alias.Name, AliasPrefix) {
+			hostDevices = append(hostDevices, device)
+		}
+	}
+	return hostDevices
+}
+
+type Registrar interface {
+	Register(callback interface{}) error
+	Deregister() error
+}
+
+func WaitHostDevicesToDetach(detacher Registrar, hostDevices []api.HostDevice, timeout time.Duration, f func() error) error {
```

Comment:
I wil try to put a proposal here to split this into the `RegisterToHostDevicesDetach` and `WaitHostDevicesToDetach` and also move the detachSRIODevices to the `sriov` package

From the caller perspect `pkg/virt-launcher/virtwrap/manager.go` it would be:

```golang
func detachHostDevices(virConn cli.Connection, dom cli.VirDomain) error {       
    domainSpec, err := util.GetDomainSpecWithFlags(dom, 0)                      
    if err != nil {                                                             
        return err                                                              
    }                                                                           
                                                                                
    if registrar := cli.NewDomainEventDeviceRemovedRegistrar(virConn, dom); registrar != nil {
        err := sriov.SafelyDetachHostDevices(domainSpec, registrar, dom)        
        if err != nil {                                                         
            return err                                                          
        }                                                                       
    }                                                                           
    return nil                                                                  
}                          
```

And from the callee `pkg/virt-launcher/virtwrap/device/sriov/hostdev.go`

```golang
func SafelyDetachHostDevices(domainSpec *api.DomainSpec, registrar Registrar, dom deviceDetacher) error {
    sriovDevices := FilterHostDevices(domainSpec)                               
                                                                                
    eventsChan, err := registerToHostDevicesDetach(registrar, sriovDevices)     
    if err != nil {                                                             
        return err                                                              
    }                                                                           
                                                                                
    err = DetachHostDevices(dom, sriovDevices)                                  
    if err != nil {                                                             
        return err                                                              
    }                                                                           
                                                                                
    return waitHostDevicesToDetach(registrar, sriovDevices, 30*time.Second, eventsChan)
}                                                                               
                                                                                
func registerToHostDevicesDetach(detacher Registrar, hostDevices []api.HostDevice) (chan string, error) {
    if len(hostDevices) == 0 {                                                  
        log.Log.Info("No host-devices to detach.")                              
        return nil, nil                                                         
    }                                                                           
                                                                                
    eventsChan := make(chan string, len(hostDevices)+2)                         
    var callback libvirt.DomainEventDeviceRemovedCallback = func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventDeviceRemoved) {
        eventsChan <- event.DevAlias                                            
    }                                                                           
                                                                                
    if err := detacher.Register(callback); err != nil {                         
        return nil, fmt.Errorf("failed to wait for host-devices detach: %v", err)
    }                                                                           
    return eventsChan, nil                                                      
}                                                                               
                                                                                
func waitHostDevicesToDetach(detacher Registrar, hostDevices []api.HostDevice, timeout time.Duration, eventsChan chan string) error {
    defer func() {                                                              
        if err := detacher.Deregister(); err != nil {                           
            log.Log.Reason(err).Errorf("failed to wait for host-devices detach, deregister failure: %v", err)
        }                                                                       
    }()                                                                         
    var detachedHostDevices []string                                            
    for {                                                                       
        select {                                                                
        case deviceAlias := <-eventsChan:                                       
            if dev := deviceLookup(hostDevices, deviceAlias); dev != nil {      
                detachedHostDevices = append(detachedHostDevices, dev.Alias.Name)
            }                                                                   
            if len(hostDevices) == len(detachedHostDevices) {                   
                return nil                                                      
            }                                                                   
        case <-time.After(timeout):                                             
            return fmt.Errorf("failed to wait for host-devices detach, timeout reached: %v", detachedHostDevices)
        }                                                                       
    }                                                                           
    return nil
}                                                                  
```

Also I see that sriov.DetachHostDevices and sriov.FilterHostDevices are only used by sriov package so they don't need to be public (I think in those cases is ok to unit test them even if they are private, at the end that's why the golang tests are living withing the package).

Reply:
Thank you for raising these points.

I have moved part of the code into the `sriov` package and amended that change.
The rest of the changes I have pushed as a separate commit, just to see if it is acceptable. If it is, I will squash it into the prev commit.

I have taken your suggestion but changed it a bit. My main problem with the original suggestion was that the registration and deregistration were in different functions, which seems inappropriate to me. And that channel was returning from a registration functions, which also looks strange.
Hope this new approach will be of your liking. Let me know what you think.